exchanges/okgroup/okgroup_websocket.go

```diff
@@ -0,0 +1,719 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// WsLogin sends a login request to websocket to enable access to authenticated endpoints
+func (o *OKGroup) WsLogin() error {
+	utcTime := time.Now().UTC()
+	unixTime := utcTime.Unix()
+	signPath := "/users/self/verify"
+	hmac := common.GetHMAC(common.HashSHA256, []byte(fmt.Sprintf("%v", unixTime)+http.MethodGet+signPath), []byte(o.APISecret))
+	base64 := common.Base64Encode(hmac)
+	resp := WebsocketEventRequest{
+		Operation: "login",
+		Arguments: []string{o.APIKey, o.ClientID, fmt.Sprintf("%v", unixTime), base64},
+	}
+	json, err := common.JSONEncode(resp)
+	if err != nil {
+		return err
+	}
+	err = o.writeToWebsocket(string(json))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// WsHandleErrorResponse sends an error message to ws handler
+func (o *OKGroup) WsHandleErrorResponse(event WebsocketErrorResponse) {
+	errorMessage := fmt.Sprintf("%v error - %v message: %s ",
+		o.GetName(), event.ErrorCode, event.Message)
+	if o.Verbose {
+		log.Error(errorMessage)
+	}
+	o.Websocket.DataHandler <- fmt.Errorf(errorMessage)
+}
+
+// GetWsChannelWithoutOrderType takes WebsocketDataResponse.Table and returns
+// The base channel name eg receive "spot/depth5:BTC-USDT" return "depth5"
+func (o *OKGroup) GetWsChannelWithoutOrderType(table string) string {
+	index := strings.Index(table, "/")
+	if index == -1 {
+		return table
+	}
+	channel := table[index+1:]
+	index = strings.Index(channel, ":")
+	// Some events do not contain a currency
+	if index == -1 {
+		return channel
+	}
+
+	return channel[:index]
+}
+
+// GetAssetTypeFromTableName gets the asset type from the table name
+// eg "spot/ticker:BTCUSD" results in "SPOT"
+func (o *OKGroup) GetAssetTypeFromTableName(table string) string {
+	assetIndex := strings.Index(table, "/")
+	return strings.ToUpper(table[:assetIndex])
+}
+
+// WsHandleDataResponse classifies the WS response and sends to appropriate handler
+func (o *OKGroup) WsHandleDataResponse(response *WebsocketDataResponse) {
+	switch o.GetWsChannelWithoutOrderType(response.Table) {
+	case okGroupWsCandle60s, okGroupWsCandle180s, okGroupWsCandle300s, okGroupWsCandle900s,
+		okGroupWsCandle1800s, okGroupWsCandle3600s, okGroupWsCandle7200s, okGroupWsCandle14400s,
+		okGroupWsCandle21600s, okGroupWsCandle43200s, okGroupWsCandle86400s, okGroupWsCandle604900s:
+		if o.Verbose {
+			log.Debugf("%v Websocket candle data received", o.GetName())
+		}
+		o.wsProcessCandles(response)
+	case okGroupWsDepth, okGroupWsDepth5:
+		if o.Verbose {
+			log.Debugf("%v Websocket orderbook data received", o.GetName())
+		}
+		// Locking, orderbooks cannot be processed out of order
+		orderbookMutex.Lock()
+		err := o.WsProcessOrderBook(response)
+		if err != nil {
+			log.Error(err)
+			subscriptionChannel := fmt.Sprintf("%v:%v", response.Table, response.Data[0].InstrumentID)
+			o.WsUnsubscribeToChannel(subscriptionChannel)
+			o.WsSubscribeToChannel(subscriptionChannel)
```

Comment:
There may be times here where if WsProcessOrderbook returns an error and then WsUnsubscribeToChannel or WsSubscribeToChannel also error that the Websocket channel subscription is never re-established
Both of these do return errors that may be work handling and reporting on

Reply:
That's true. I'll handle the errors, but if it fails, even after x attempts to unsubscribe, resubscribe, what do you think should be done? Another log.Error highlighting the failure?

Reply:
It comes down to:

Can this exchange still operate and trade without issues if this happens?
If not can the exchange be disabled without impacting anything further in the trade engine?

If it can't and will impact the trading these are times where even a fatal log might be needed